### ISSUE TITLE
mention css/style injection in 14.4.3

### ISSUE THREAD

**Issue by RafaelGreen1** [OWASP/ASVS#1286]
I would mention css/style injection as part of the section 14.4.3 (HTTP security headers -> Content Security Policy (CSP))

**Comment by elarlang** [comment]
Current requirement:

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **14.4.3** | Verify that a Content Security Policy (CSP) response header is in place that helps mitigate impact for XSS attacks like HTML, DOM, JSON, and JavaScript injection vulnerabilities. | ✓ | ✓ | ✓ | 1021 |

How often you have seen CSS injection? In general I think "CSS injection" case is covered at the moment with "mitigate impact for XSS attacks" and personally I don't think it needs extra attention in this "like this and that" list.

**Comment by RafaelGreen1** [comment]
I have seen CSS injection in few security tutorials (one of them is https://learn.snyk.io/lessons/xss/javascript/)
From what I understand, it's very important to protect against this attack because the attacker can change the whole view of the website via styling (and actually show on screen whatever they want).
I just thought it might be good idea to add it to the list following the word "like". like this:

> Verify that a Content Security Policy (CSP) response header is in place that helps mitigate impact for XSS attacks like HTML, DOM, **CSS**, JSON, and JavaScript injection vulnerabilities.

**Comment by elarlang** [comment]
My question was more for "have you seen CSS injection in practice", personally I can recall maybe few cases per 10 years and this is the reason why I say, that the case itself is covered with bubble name "XSS attacks".

And to be clear - `Content-Security-Policy` is second layer defence - it tries to mitigate the impact if the problem already exists. 

If you say, that `style-src` should mitigate CSS injection, it most likely means, that HTML document contains `<style>` tag and manipulatable userinput is presented in CSS syntax. In this case strict `style-src` value can not help you anymore - you can limit loading external CSS, but you can manipulate CSS in HTML document directly.

For actual mitigation you should look to requirements like:
* V5.3.1 Verify that output encoding is relevant for the interpreter and context required. For example, use encoders specifically for HTML values, HTML attributes, JavaScript, URL parameters, HTTP headers, SMTP, and others as the context requires, especially from untrusted inputs (e.g. names with Unicode or apostrophes, such as ねこ or O'Hara).
* V5.3.3 Verify that context-aware, preferably automated - or at worst, manual - output escaping protects against reflected, stored, and DOM based XSS.

All mentioned requirements can be improved, but I can not see the need to put CSS injection to special spotlight as it happens rarely enough and it is covered with bubble name "XSS".

I leave issue opened for other comments.

**Comment by willfarrell** [comment]
This reminds me of a post @ScottHelme wrote recently, [`Can you get pwned with CSS?`](https://scotthelme.co.uk/can-you-get-pwned-with-css/), when considering changing the rating for https://securityheaders.com to allow `style-src 'self' 'unsafe-inline';` within CSP for an A+ rating. 

> With popular frameworks like Angular [requiring](https://angular.io/guide/security#content-security-policy) 'unsafe-inline' in the style-src, it may even be something that's totally out of your control.

Inlining css for elements above-the-fold is a known strategy for improving load time performance.

**Comment by willfarrell** [comment]
@tghosth It can be addressed using `style-src` without including `unsafe-inline` in the Content Security Policy. 

Including CSS XSS within ASVS will, hopefully, push the browsers to include a way to inline css securely (i.e. `<style integrity="...">...`).

**Comment by elarlang** [comment]
> Including CSS XSS within ASVS will, hopefully, push the browsers to include a way to inline css securely (i.e. `<style integrity="...">...`).

Related issue(s) are discussed in: https://github.com/OWASP/ASVS/issues/1293

**Comment by ScottHelme** [comment]
@willfarrell by 'CSS XSS' do you mean CSS-injection or have you seen more powerful XSS-like attacks with CSS? I think there's a big difference between the two and I'd like to check if I'm understanding properly!

As for a way to include inline CSS securely, CSP 2 already provides for inline style tags: https://www.w3.org/TR/CSP2/#source-list-valid-hashes

```
Content-Security-Policy: style-src 'sha256-RB20JxKPtBo78ZjWkZ+GR4yOncuhVeK20jxJPz4x86c='

<html>
    <style>
         color: #e6400c;
    </style>
</html>
```

CSP 3 goes a little further and allows you to distinguish between `style-src-elem` and `style-src-attr` which breaks apart `<style>` tags and `style=""` attributes: https://www.w3.org/TR/CSP3/#directive-style-src-attr

**Comment by willfarrell** [comment]
@ScottHelme Yes, I was referring to CSS-injection (comment updated). Thanks for sharing that example, I wasn't aware one could CSP an integrity hash to an inline style, will be testing out for sure.

**Comment by elarlang** [comment]
CSS is now covered in V5.3.1 and V14.4.3 (see https://github.com/OWASP/ASVS/pull/1296/commits/96f3e6a8c34654c99ce7134ea0f22b99c9617250) and will be covered in V10.3.2 via #1293.

Closing this one.
